### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by lima-limon-inc**
- Implements all the LC3 Operands
- Includes a debug mode
- Includes a LC3 compiler as a submodule to make compiling LC3 asm easier.

### CODE REVIEW

File: src/virtual_machine.rs [hunk 1]
```diff
@@ -0,0 +1,900 @@
...
+            debug,
+        }
+    }
+
+    fn load_instruction(&mut self) -> Result<u16, VMError> {
+        self.memory_read(self.rpc)
+    }
+
+    // QUESTION: Maybe implement from?
+    fn u16_to_char(tochar: u16) -> Result<char, VMError> {
```
Comment (juanbono) [inline src/virtual_machine.rs]: I think there is a way of doing this in the standard library.

File: src/virtual_machine.rs [hunk 2]
```diff
@@ -0,0 +1,900 @@
...
+
+        let addr = addr as usize;
+        if let Some(value) = self.memory.get(addr) {
+            Ok(*value)
+        } else {
+            Err(VMError::OutOfRangeError)
+        }
+    }
+
+    fn decode_instruction(&self, binary_repr: u16) -> Result<Opcode, VMError> {
```
Comment (juanbono) [inline src/virtual_machine.rs]: This should be an Opcode's method instead of a VM's method.

**Review by juanbono — CHANGES_REQUESTED** [review]
See comments.

Also you can split the virtual_machine.rs file into 2:

- One containing vm related structs and methods.
- One with the Opcode struct and related methods and structs.

**Comment by lima-limon-inc** [comment]
On it! Thanks bono

**Pull request merged**